### PR TITLE
Add exclusion for tables of type Setup for Rule LC0023

### PR DESCRIPTION
**Pull request by Arthurvdv**
This will resolve issue https://github.com/StefanMaron/BusinessCentral.LinterCop/issues/241.

Because of the difficulty to determine if a table is a (Singleton) Setup table, I think the criteria should be;

* The first field of the Table should have the exact (case sensitive) name "Primary Key"
* The first field of the Table should be of type Code (can be Code[10], Code[20], the length can be different)

### CODE REVIEW

File: Design/Rule0023AlwaysSpecifyFieldgroups.cs [hunk 1]
```diff
@@ -20,6 +21,10 @@ private void CheckFieldgroups(SymbolAnalysisContext ctx)
             try
             {
                 ITableTypeSymbol table = (ITableTypeSymbol)ctx.Symbol;
+                if (table.Length > 0)
```
Comment (StefanMaron) [inline Design/Rule0023AlwaysSpecifyFieldgroups.cs:24]: Under which condition is table length not more than 0? What does this condition catch?

File: Design/Rule0023AlwaysSpecifyFieldgroups.cs [hunk 2]
```diff
@@ -20,6 +21,10 @@ private void CheckFieldgroups(SymbolAnalysisContext ctx)
...
+                {
+                    if (IsTableOfTypeSetupTable(table.Fields[0])) return;
```
Comment (StefanMaron) [inline Design/Rule0023AlwaysSpecifyFieldgroups.cs:26]: I would consider to pass in the whole table here...

File: Design/Rule0023AlwaysSpecifyFieldgroups.cs [hunk 3]
```diff
@@ -40,5 +45,11 @@ private void CheckFieldgroups(SymbolAnalysisContext ctx)
                 ctx.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptors.Rule0000ErrorInRule, ctx.Symbol.GetLocation(), new Object[] { "Rule0023", "ArgumentOutOfRangeException", "" }));
             }
         }
+
+        private static bool IsTableOfTypeSetupTable(IFieldSymbol field)
+        {
+            // The first field of the table should be of type Code and exactly (case sensitive) called 'Primary Key'
+            return (field.GetTypeSymbol().NavTypeKind == NavTypeKind.Code && field.Name == "Primary Key");
```
Comment (StefanMaron) [inline Design/Rule0023AlwaysSpecifyFieldgroups.cs:52]: ... so you can also check if the table only has a single field as PK, just to be sure

**Review by StefanMaron — CHANGES_REQUESTED** [review]

**Comment by ShareTEC** [comment]
Checking that the first field is "Primary Key" and that indeed the "Primary Key" field is the one and only key in this table is a valid criteria. Something I would certainly check on. A "cue" table is then also covered.

If you look at the guidelines of all "setup" tables in BC, then you can also see that the name of a setup tables ends on "Setup". Maybe this can be used as a second criteria? But that would bypass "cue" tables.

**Comment by StefanMaron** [comment]
I would not check the name TBH. Since the name is limited to 30 characters and 3-5 chars affix is needed, the Setup keyword could be dropped.

Another Idea would be to introduce a Comment line with specific wording. Like `// Setup Table` to indicate that this is setup. Since we dont have a `TableType = Setup`

**Comment by ShareTEC** [comment]
Yes, I was also thinking about a comment line... I don't mind that solution.

File: Design/Rule0023AlwaysSpecifyFieldgroups.cs (hunk 1 above)
Comment (Arthurvdv) [reply Design/Rule0023AlwaysSpecifyFieldgroups.cs]: Oops, this is a typo and should be table.Fields.Length. Fixed with the change to the PrimaryKey (table.PrimaryKey.Fields.Length != 1)

File: Design/Rule0023AlwaysSpecifyFieldgroups.cs (hunk 2 above)
Comment (Arthurvdv) [reply Design/Rule0023AlwaysSpecifyFieldgroups.cs]: Good idea, done.

**Comment by Arthurvdv** [comment]
Updated the logic to look at the Primary Key and retrieve the field specifications from this. This makes the rule a bit more robust when, for some reason, the Primary Field is not declared as the first field it still works.

I'm having doubt on the comment line. Is adding the comment the same effort as adding a pragma? Or is this a two-stage validation step, where first look for the keyword 'Setup' in the name of the table. And, if not present, fallback on searching on a comment line which contains the keyword 'Setup'.

**Review by StefanMaron — APPROVED** [review]

**Review by StefanMaron — APPROVED** [review]

**Pull request merged**